### PR TITLE
Av sync rtcp

### PR DESCRIPTION
**Pull request by kekkokk**
Implemented a simple version of rtcp a/v sync for the recorder.

I'm new to C, so I need someone to help me find bugs. this implementation seems working, but since chrome will choose random start rtp timestamp, I don't know how it reacts to overflow or negative rtp. 
In 2 days testing found no problems.

Tested with pcmu / opus

### CODE REVIEW

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 1]
```diff
@@ -398,9 +459,16 @@ void ExternalOutput::queueData(char* buffer, int length, packetType type){
           context_->oformat->audio_codec = AV_CODEC_ID_PCM_MULAW;
         }
     }
-    if (needToSendFir_ && videoSourceSsrc_) {
+
+    timeval time;
+    gettimeofday(&time, NULL);
```
Comment (jnoring) [inline erizo/src/erizo/media/ExternalOutput.cpp]: You'll want to use a monotonic timer here; I realize other parts of the code don't (including parts I use) but it's more stable.

```
uint64_t getMonotonicTimeInUSec() {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (ts.tv_sec * uint64_t(1000000)) + (ts.tv_nsec / uint64_t(1000));
}
```

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 2]
```diff
@@ -301,7 +343,13 @@ void ExternalOutput::writeVideoData(char* buf, int len){
         avpkt.size = unpackagedSize_;
         avpkt.pts = timestampToWrite;
         avpkt.stream_index = 0;
-        av_interleaved_write_frame(context_, &avpkt);   // takes ownership of the packet
+        if (av_interleaved_write_frame(context_, &avpkt) < 0) {
+                ELOG_ERROR("Write video frame failed");
```
Comment (jnoring) [inline erizo/src/erizo/media/ExternalOutput.cpp]: Would be nice to print out the error received here

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 3]
```diff
@@ -163,7 +180,13 @@ void ExternalOutput::writeAudioData(char* buf, int len){
     avpkt.size = len - head->getHeaderLength();
     avpkt.pts = timestampToWrite;
     avpkt.stream_index = 1;
-    av_interleaved_write_frame(context_, &avpkt);   // takes ownership of the packet
+    if (av_interleaved_write_frame(context_, &avpkt) < 0) {
+        ELOG_ERROR("Write audio frame failed");
```
Comment (jnoring) [inline erizo/src/erizo/media/ExternalOutput.cpp]: Would be nice to print out the error received here

File: erizo/src/erizo/media/ExternalOutput.cpp [hunk 4]
```diff
@@ -315,7 +363,11 @@ int ExternalOutput::deliverAudioData_(char* buf, int len) {
 int ExternalOutput::deliverVideoData_(char* buf, int len) {
     if (videoSourceSsrc_ == 0){
       RtpHeader* h = reinterpret_cast<RtpHeader*>(buf);
-      videoSourceSsrc_ = h->getSSRC();
+      if (h->getPayloadType() == RED_90000_PT) {
```
Comment (jnoring) [inline erizo/src/erizo/media/ExternalOutput.cpp]: I'd turn off FEC in your SDP for any number of reasons.  I've completely abandoned trying to use forward error correct, which removes the RED headers entirely.

Comment (jnoring) [inline erizo/src/erizo/media/ExternalOutput.cpp]: If you do turn it off, you'll also need a check to see if the payload type is VP8/VP9 or some other video codec

File: erizo/src/erizo/rtp/RtpHeaders.h [hunk 5]
```diff
@@ -420,9 +420,18 @@ namespace erizo{
       }
       inline uint32_t getOctetsSent(){
         return ntohl(report.senderReport.octetssent);
-      }      
+      }
       inline uint64_t getNtpTimestamp(){
-       return (((uint64_t)htonl(report.senderReport.ntptimestamp)) << 32) + htonl(report.senderReport.ntptimestamp >> 32);
+        return (((uint64_t)htonl(report.senderReport.ntptimestamp)) << 32) + htonl(report.senderReport.ntptimestamp >> 32);
```
Comment (jnoring) [inline erizo/src/erizo/rtp/RtpHeaders.h]: shouldn't this be `ntohl(report.senderReport.ntptimestamp)` ?

File: erizo/src/erizo/rtp/RtpHeaders.h [hunk 6]
```diff
@@ -420,9 +420,18 @@ namespace erizo{
...
+      }
+      inline uint32_t getNtpTimestampMSW() {
+        return htonl((uint32_t)report.senderReport.ntptimestamp);
```
Comment (jnoring) [inline erizo/src/erizo/rtp/RtpHeaders.h]: I think you want `ntohl`

File: erizo/src/erizo/rtp/RtpHeaders.h [hunk 7]
```diff
@@ -420,9 +420,18 @@ namespace erizo{
...
+      }
+      inline uint32_t getNtpTimestampLSW() {
+        return htonl((uint32_t)(report.senderReport.ntptimestamp >> 32));
```
Comment (jnoring) [inline erizo/src/erizo/rtp/RtpHeaders.h]: I think you want `ntohl`

File: erizo/src/erizo/media/SyncUtils.cpp [hunk 8]
```diff
@@ -0,0 +1,108 @@
+
```
Comment (jnoring) [inline erizo/src/erizo/media/SyncUtils.cpp]: This looks awesome, but I'd consider adding some unit tests (can be found under `erizo/src/test/`)

File: erizo/src/erizo/rtp/RtpHeaders.h (hunk 5 above)
Comment (kekkokk) [inline erizo/src/erizo/rtp/RtpHeaders.h]: @lodoyun could you please check this? This code was already committed in the current master branch.

File: erizo/src/erizo/media/ExternalOutput.cpp (hunk 4 above)
Comment (kekkokk) [inline erizo/src/erizo/media/ExternalOutput.cpp]: this is a precious information you gave me. Thanks a lot. Tomorrow morning i'll test without fec

File: erizo/src/erizo/rtp/RtpHeaders.h (hunk 6 above)
Comment (kekkokk) [inline erizo/src/erizo/rtp/RtpHeaders.h]: Yeah, this is related to the code which was already in rtpheaders, which uses

return (((uint64_t)htonl(report.senderReport.ntptimestamp)) << 32) + htonl(report.senderReport.ntptimestamp >> 32);

I'll investigate. thanks

Comment (lodoyun) [inline erizo/src/erizo/rtp/RtpHeaders.h]: Same here, in general, should be ntoh for all _gets_

File: erizo/src/erizo/media/ExternalOutput.cpp (hunk 1 above)
Comment (lodoyun) [inline erizo/src/erizo/media/ExternalOutput.cpp]: You're right. we'll do this Erizo-wide in the near future

**Comment by kekkokk** [comment]
No probl, 
I made a better implementation in this PR https://github.com/ging/licode/pull/549

Unfortunately, as I wrote in the PR, I'm absolutely new to C and it's my first piece of code in this language. I really don't know where to start to make some unit tests, and currently my job schedule doesn't allow me to take some time to implement.

I'm glad to share my code and ideas, if anyone could help me with tests I'll be very glad. 
Otherwise, I'll leave the PR opened, and hope in the future to be able to do these tests.

I'll close this one since it's quite obsolete

**Pull request closed without merging**